src/chunkserver/chunkserver-common/hdd_utils.cc

```diff
@@ -102,3 +107,328 @@ void hddRemoveChunkFromContainers(IChunk *chunk) {
+		}
+
+		return chunk;
+	}
+
+	chunk = chunkIter->second.get();
+	effectiveDisk = chunk->owner();
+
+	if (creationMode == disk::ChunkGetMode::kCreateOnly) {
+		if (chunk->state() == ChunkState::Available ||
+		    chunk->state() == ChunkState::Locked) {
+			return nullptr;
+		}
+	}
+
+	while (true) {
+		switch (chunk->state()) {
+		case ChunkState::Available:
+			chunk->setState(ChunkState::Locked);
+			chunksMapLock.unlock();
+			if (chunk->validAttr() == 0) {
+				if (effectiveDisk->updateChunkAttributes(chunk, false) ==
+				    SAUNAFS_ERROR_NOCHUNK) {
+					// The chunk was found as available, but we can not
+					// update its attributes, let's recreate it only if
+					// requested
+					if (creationMode != disk::ChunkGetMode::kFindOnly) {
+						effectiveDisk->unlinkChunk(chunk);
+						chunksMapLock.lock();
+						chunk = hddRecreateChunk(effectiveDisk, chunk, chunkid,
+						                         chunkType);
+						return chunk;
+					}
+
+					// The Chunk is damaged, remove it from disk and from
+					// memory
+					hddReportDamagedChunk(chunk->id(), chunk->type());
+					effectiveDisk->unlinkChunk(chunk);
+					hddDeleteChunkFromRegistry(chunk);
+					return nullptr;
+				}
+			}
+			return chunk;
+		case ChunkState::Deleted:
+			if (creationMode != disk::ChunkGetMode::kFindOnly) {  // Reuse it
+				chunk =
+				    hddRecreateChunk(effectiveDisk, chunk, chunkid, chunkType);
+				return chunk;
+			}
+			if (chunk->condVar() !=
+			    nullptr) {  // waiting threads - wake them up
+				chunk->condVar()->condVar.notify_one();
+			} else {  // no more waiting threads - remove
+				hddRemoveChunkFromContainers(chunk);
+			}
+			return nullptr;
+		case ChunkState::ToBeDeleted:
+		case ChunkState::Locked:
+			if (chunk->condVar() == nullptr) {
+				// Try to reuse one if possible.
+				if (!gFreeCondVars.empty()) {
+					chunk->setCondVar(std::move(gFreeCondVars.back()));
+					gFreeCondVars.pop_back();
+				} else {
+					chunk->setCondVar(std::make_unique<CondVarWithWaitCount>());
+				}
+			}
+			chunk->condVar()->numberOfWaitingThreads++;
+			auto status = chunk->condVar()->condVar.wait_for(
+			    chunksMapLock,
+			    std::chrono::seconds(kSecondsToWaitForLockedChunk_));
+			chunk->condVar()->numberOfWaitingThreads--;
+			if (chunk->condVar()->numberOfWaitingThreads == 0) {
+				// No more waiting threads, store it to be reused
+				gFreeCondVars.emplace_back(std::move(chunk->condVar()));
+			}
+			if (status == std::cv_status::timeout) {
+				safs_pretty_syslog(LOG_WARNING, "Chunk locked for long time");
+				return nullptr;
+			}
+		}
+	}
+}
+
+IChunk *hddRecreateChunk(IDisk *disk, IChunk *chunk, uint64_t chunkId,
+                         ChunkPartType type) {
+	std::unique_ptr<CondVarWithWaitCount> waiting;
+
+	if (chunk != ChunkNotFound) {
+		assert(chunk->id() == chunkId);
+
+		if (chunk->state() != ChunkState::Deleted &&
+		    chunk->owner() != nullptr) {
```

Comment:
Consider comparing `chunk->owner()` with `DiskNotFound` instead of `nullptr`. 
This is only a suggestion because is out of the scope of this PR.